Contracts & Harnesses for `NonNull::read`, `NonNull::read_volatile`, `NonNull::read_unaligned`

QinyuanWu opened the pull request:
Towards #52 

Contracts & Harnesses for `NonNull::read`, `NonNull::read_volatile`, `NonNull::read_unaligned`
### Discussion
For the ensures clause in the contract for `NonNull::read`, I could not directly compare the value pointed by self and result due to unsized and lack of PartialEq implementation, and the comparison is currently moved to the harness.
### Verification
Successful verification:
```
SUMMARY:
 ** 0 of 141 failed

VERIFICATION:- SUCCESSFUL
Verification Time: 0.1760525s

Complete - 5 successfully verified harnesses, 0 failures, 5 total.
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

zhassan-aws commented on library/core/src/ptr/non_null.rs:
```diff
@@ -898,12 +898,16 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(self.pointer.is_aligned() && ub_checks::can_dereference(self.pointer))]
```
`can_dereference` already checks that the pointer is aligned.

zhassan-aws commented on library/core/src/ptr/non_null.rs:
```diff
@@ -898,12 +898,16 @@ impl<T: ?Sized> NonNull<T> {
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(self.pointer.is_aligned() && ub_checks::can_dereference(self.pointer))]
+    #[ensures(|result| self.pointer.is_aligned())] // unsafe { *result == *self.pointer } can't be used due to unsized and lack of PartialEq implementation, moved to harness
```
You can remove the `ensures` clause. Since there is no `modifies` clause in this contract, Kani will check that the value is not modified.

zhassan-aws commented on library/core/src/ptr/non_null.rs:
```diff
@@ -925,6 +929,8 @@ impl<T: ?Sized> NonNull<T> {
     #[inline]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(self.pointer.is_aligned() && ub_checks::can_dereference(self.pointer))]
```
Remove `is_aligned`.

zhassan-aws commented on library/core/src/ptr/non_null.rs:
```diff
@@ -925,6 +929,8 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(self.pointer.is_aligned() && ub_checks::can_dereference(self.pointer))]
+    #[ensures(|result| self.pointer.is_aligned())]
```
Drop the `ensures`.

zhassan-aws commented on library/core/src/ptr/non_null.rs:
```diff
@@ -945,6 +951,7 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(ub_checks::can_dereference(self.pointer))]
```
Use `can_read_unaligned`

zhassan-aws commented on library/core/src/ptr/non_null.rs:
```diff
@@ -1829,4 +1836,83 @@ mod verify {
+        // array example
+        const arr_len: usize = 10;
+        let offset = kani::any_where(|x| * x < arr_len);
+        kani::assume(offset >= 0);
```
This assumption is not needed.

zhassan-aws commented on library/core/src/ptr/non_null.rs:
```diff
@@ -1829,4 +1836,83 @@ mod verify {
+        }
+
+        // array example
+        const arr_len: usize = 10;
```
Use upper case for constants.

zhassan-aws commented on library/core/src/ptr/non_null.rs:
```diff
@@ -1829,4 +1836,83 @@ mod verify {
+        kani::assume(offset >= 0);
+        let arr: [i8; arr_len] = kani::any();
+        let raw_ptr: *mut i8 = arr.as_ptr() as *mut i8;  
+        let nonnull_ptr = unsafe { NonNull::new(raw_ptr).unwrap()};
```
The `PointerGenerator` should be more suitable for usage here.

QinyuanWu commented:
Addressed previous PR comments and Increased SIZE:
```
Checking harness ptr::non_null::verify::non_null_check_read_unaligned...
SUMMARY:
 ** 0 of 261 failed (4 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 18.91408s

Checking harness ptr::non_null::verify::non_null_check_read_volatile...
SUMMARY:
 ** 0 of 237 failed (4 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 15.834378s
Checking harness ptr::non_null::verify::non_null_check_read...
SUMMARY:
 ** 0 of 229 failed (4 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 38.454617s

Complete - 3 successfully verified harnesses, 0 failures, 3 total.
```

QinyuanWu commented:
@zhassan-aws Thank you! Waiting for another review.

zhassan-aws reviewed (approved):
Can you please resolve the conflict?

celinval commented on library/core/src/ptr/non_null.rs:
```diff
@@ -1829,4 +1833,90 @@ mod verify {
+        let nonnull_ptr = NonNull::new(unaligned_ptr as *mut usize).unwrap();
+        unsafe { 
+            let result = nonnull_ptr.read_unaligned();
+            
```
FYI, the PointerGenerator will generate unaligned pointers.

QinyuanWu replied:
Thanks for the suggestion! Changed to PointerGenerator.

QinyuanWu replied:
Thanks for the suggestion! Changed to `PointerGenerator`.

celinval commented on library/core/src/ptr/non_null.rs:
```diff
@@ -1829,4 +1833,90 @@ mod verify {
+        unsafe { 
+            let result = nonnull_ptr.read_unaligned();
+            
+            let raw_result_ptr = &result as *const usize as *const u8;
+            let raw_original_ptr = &arr[offset] as *const u8;
+            for i in 0..core::mem::size_of::<usize>() {
+                unsafe {
+                    // Dereference and compare each byte
+                    assert_eq!(*raw_original_ptr.add(i), *raw_result_ptr.add(i));
+                }
```
I don't think we need this

QinyuanWu replied:
@celinval should we check whether `read_unaligned` reads the value correctly?

Eh2406 commented:
I've then asked to help review this.

HuStmpHrrr commented:
LGTM

QinyuanWu commented:
@zhassan-aws resolved conflicts and addressed all comments, ready to merge:
```
VERIFICATION:- SUCCESSFUL
Verification Time: 0.173812s

Complete - 9 successfully verified harnesses, 0 failures, 9 total.
```

zhassan-aws commented:
trigger approval workflow

zhassan-aws commented:
trigger approval workflow
